scripts/index.py

```diff
@@ -57,63 +48,95 @@ def print_progress(counter, start_time):
         })
 
 
-class ServiceIndexer(object):
-    def __init__(self, endpoint, access_token, index):
-        self.endpoint = endpoint
-        self.access_token = access_token
+class IndexerBase(object):
+    def __init__(self, document_type, data_client, search_client, index):
+        self.document_type = document_type
         self.index = index
-
-    def __call__(self, service):
-        client = dmapiclient.SearchAPIClient(self.endpoint, self.access_token)
-
-        try:
-            self.index_service(client, service)
-            return True
-        except dmapiclient.APIError:
-            logger.exception("{service_id} not indexed", extra={'service_id': service.get('id')})
-            return False
-
-    @backoff.on_exception(backoff.expo, dmapiclient.HTTPError, max_tries=5)
-    def index_service(self, client, service):
-        if service['status'] == 'published':
-            client.index(self.index, service['id'], service)
-        else:
-            client.delete(self.index, service['id'])
+        self.search_client = search_client
+        self.data_client = data_client
 
     def create_index(self, mapping):
-        client = dmapiclient.SearchAPIClient(self.endpoint, self.access_token)
         logger.info("Creating {index} index", extra={'index': self.index})
 
         try:
-            result = client.create_index(self.index, mapping=mapping)
+            result = self.search_client.create_index(self.index, mapping=mapping)
             logger.info("Index creation response: {response}", extra={'response': result})
         except dmapiclient.HTTPError as e:
             if 'already exists as alias' in str(e.message):
                 logger.info("Skipping index creation for alias {index}", extra={'index': self.index})
             else:
                 raise
 
+    def request_items(self, frameworks):
```

Comment:
For my own understanding, why do we need to define this here if it's defined in both the sub classes? Is it an OOP thing?

Reply:
This is just a way of declaring that it's a required part of the interface. There are other ways of doing that too, but this often seems the easiest. It's a way of documenting that if you are going to make a new subclass, then you have to have this method, is all.